R/as_mapper.R

```diff
@@ -62,35 +61,77 @@ as_function <- function(...) {
 
 #' Pluck out a single an element from a vector or environment
 #'
-#' This is a generalised form of `[[` which allows you to index by
-#' name, position, or attribute.
+#' @description
+#'
+#' This is a generalised form of `[[` which allows you to index deeply
+#' and flexibly into data structures. It supports R standard accessors
+#' like integer positions and string names and accepts any accessor
+#' functions.
+#'
+#' `pluck()` is often more readable than a mix of operators and
+#' accessors because it reads linearly and is free of syntactic
+#' cruft. Compare: `accessor(x[[1]])$foo` to `pluck(x, 1, accessor,
+#' "foo")`.
 #'
-#' @param x A vector or environment
-#' @param index A list indexing into the object
-#' @param default Value to use if target is empty or absent.
+#' @param .x A vector or environment
+#' @param ... A list of accessors for indexing into the object. Can be
+#'   an integer position, a string name, or an accessor function.
+#'   These dots [splice lists automatically][rlang::dots_splice]. This
+#'   means you can supply arguments and lists of arguments
+#'   indistinctly.
+#' @param .default Value to use if target is empty or absent.
 #' @keywords internal
 #' @export
-pluck <- function(x, index, default = NULL) {
-  .Call(extract_impl, x, index, default)
+#' @examples
+#' # pluck() supports integer positions, string names, and functions.
+#' # Using functions, you can easily extend pluck(). Let's create a
+#' # list of data structures:
+#' obj1 <- list("a", list(1, elt = "foobar"))
+#' obj2 <- list("b", list(2, elt = "foobaz"))
+#' x <- list(obj1, obj2)
+#'
+#' # And now an accessor for these complex data structures:
+#' my_element <- function(x) x[[2]]$elt
+#'
+#' # The accessor can then be passed to pluck:
+#' pluck(x, 1, my_element)
+#' pluck(x, 2, my_element)
+#'
+#' # Even for this simple data structure, this is more readable than
+#' # the alternative form:
```

Comment:
because you have to read both from right-to-left and left-to-right in different parts of the code